### PR TITLE
WebSocket Client Support

### LINKED ISSUE #650: WebSocket client
**Is your feature request related to a problem? Please describe.**
I want to be able to use the http client for websocket connections

**Describe the solution you'd like**
We can start with a basic API that looks like this. Later on, add support for client configuration or potentially support `SocketApp` directly.

```scala
val fooBar = Socket.collect[WebSocketFrame] {
  case WebSocketFrame.Text("FOO") => ZStream.succeed(WebSocketFrame.text("BAR"))
  case WebSocketFrame.Text("BAR") => ZStream.succeed(WebSocketFrame.text("FOO"))
}

Client.socket(fooBar) // Task[Unit]
```

**Describe alternatives you've considered**
- None

### PR DESCRIPTION
**Pull request by tusharmath**
Closes #650 

Depends on #894 #898 #900 #901 #902 #903

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/http/Request.scala [hunk 1]
```diff
@@ -98,10 +98,16 @@ object Request {
     remoteAddress: Option[InetAddress] = None,
     data: HttpData = HttpData.Empty,
   ): Request = {
-    val m  = method
-    val u  = url
-    val h  = headers
-    val ra = remoteAddress
+    val hostHeaders = (url.host, url.port) match {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Request.scala]: I don't think this is required.

Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Request.scala]: I don't think this is required.

File: zio-http/src/test/scala/zhttp/http/RequestSpec.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package zhttp.http
+
+import zio.test.Assertion.{equalTo, isSome}
+import zio.test.{DefaultRunnableSpec, _}
+
+object RequestSpec extends DefaultRunnableSpec {
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/RequestSpec.scala]: Delete RequestSpec

File: zio-http/src/test/scala/zhttp/internal/HttpGen.scala [hunk 3]
```diff
@@ -93,6 +96,10 @@ object HttpGen {
     ),
   )
 
+  def scheme: Gen[Any, Scheme]                    = Gen.fromIterable(List(HTTP, HTTPS, WS, WSS))
+  def jHttpScheme: Gen[Any, HttpScheme]           = Gen.fromIterable(List(HttpScheme.HTTP, HttpScheme.HTTPS))
+  def jWebSocketScheme: Gen[Any, WebSocketScheme] = Gen.fromIterable(List(WebSocketScheme.WS, WebSocketScheme.WSS))
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/internal/HttpGen.scala]: Let's move the netty gens close to the tests and move it out of this file. This file is only for ZIO Http Gens.

File: zio-http/src/test/scala/zhttp/service/WebSocketServerSpec.scala [hunk 4]
```diff
@@ -13,21 +13,31 @@ import zio.test._
...
+              .map(_.status)
+              .catchAll {
+                case None        => ZIO.fail(new Error("No status code"))
+                case Some(error) => ZIO.fail(error)
+              }
+          } yield code == Status.SWITCHING_PROTOCOLS
+        }
+
+        val allTrue = codes.map(_.count(identity))
+        assertM(allTrue)(equalTo(1024))
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/WebSocketServerSpec.scala]: Simplify code like it was before.

File: zio-http/src/main/scala/zhttp/http/Scheme.scala [hunk 5]
```diff
@@ -1,29 +1,70 @@
...
+    case _            => false
+  }
+
+  def toJHttpScheme: Option[HttpScheme] = self match {
+    case HTTP  => Option(HttpScheme.HTTP)
+    case HTTPS => Option(HttpScheme.HTTPS)
+    case _     => None
+  }
+
+  def toWebSocketScheme: Option[WebSocketScheme] = self match {
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/http/Scheme.scala]:
```suggestion
  def toJWebSocketScheme: Option[WebSocketScheme] = self match {
```
for consistency in nomenclature

**Review by amitksingh1490 — APPROVED** [review]

File: zio-http/src/test/scala/zhttp/service/WebSocketServerSpec.scala [hunk 6]
```diff
@@ -13,21 +13,22 @@ import zio.test._
...
   def websocketSpec = suite("WebSocket Server") {
     suite("connections") {
       testM("Multiple websocket upgrades") {
-        val response = Socket.succeed(WebSocketFrame.text("BAR")).toResponse
-        val app      = Http.fromZIO(response)
-        assertM(app.deployWebSocket.map(_.code.code).run(path = !! / "subscriptions").repeatN(1024))(equalTo(101))
+        val app   = Socket.succeed(WebSocketFrame.text("BAR")).toHttp.deployWS
+        val codes = ZIO
+          .foreach(1 to 1024)(_ => app(Socket.empty.toSocketApp).map(_.status))
+          .map(r => r.count(_ == Status.SWITCHING_PROTOCOLS))
```
Comment (girdharshubham) [inline zio-http/src/test/scala/zhttp/service/WebSocketServerSpec.scala]:
```suggestion
          .map(_.count(_ == Status.SWITCHING_PROTOCOLS))
```

**Review by girdharshubham — CHANGES_REQUESTED** [review]

File: zio-http/src/main/scala/zhttp/service/Client.scala [hunk 7]
```diff
@@ -1,167 +1,181 @@
...
+      val isSSL       = req.url.scheme.exists(_.isSecure)
+
+      val initializer = new ChannelInitializer[Channel]() {
+        override def initChannel(ch: Channel): Unit = {
+
+          val pipeline                    = ch.pipeline()
+          val sslOption: ClientSSLOptions = req.attribute.ssl.getOrElse(ClientSSLOptions.DefaultSSL)
+
+          // If a https or wss request is made we need to add the ssl handler at the starting of the pipeline.
+          if (isSSL) pipeline.addLast("SSLHandler", ClientSSLHandler.ssl(sslOption).newHandler(ch.alloc))
```
Comment (girdharshubham) [inline zio-http/src/main/scala/zhttp/service/Client.scala]: Pick it from the package object

File: zio-http/src/main/scala/zhttp/http/Scheme.scala (hunk 5 above)
Comment (girdharshubham) [reply zio-http/src/main/scala/zhttp/http/Scheme.scala]: This is done. Thanks, Amit.

**Review by girdharshubham — APPROVED** [review]

**Review by amitksingh1490 — APPROVED** [review]

**Pull request merged**